File: Floor in BST - GFG/floor-in-bst.java
//{ Driver Code Starts
// Initial Template for Java

import java.util.*;
import java.lang.*;
import java.math.*;
import java.io.*;

class Node {
    int data;
    Node right;
    Node left;

    Node(int val) {
        data = val;
        right = null;
        left = null;
    }
}

class GFG {

    public static Node insert(Node tree, int val) {
        Node temp = null;
        if (tree == null) {
            return new Node(val);
        }

        if (val < tree.data) {
            tree.left = insert(tree.left, val);
        } else if (val > tree.data) {
            tree.right = insert(tree.right, val);
        }

        return tree;
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        while (T-- > 0) {
            Node root = null;
            int n = sc.nextInt();
            for (int i = 0; i < n; i++) {
                int k = sc.nextInt();
                root = insert(root, k);
            }

            int s = sc.nextInt();

            Solution obj = new Solution();
            int ans = obj.floor(root, s);
            System.out.println(ans);
        }
    }
}

// } Driver Code Ends


// User function Template for Java

class Solution {
    public static int floor(Node root, int x) {
        int fl = -1; // variable to store the floor value

        while (root != null) { // loop until root is null
            if (root.data == x) { // check if root value is equal to x
                return root.data; // return the root value as it is the floor value
            }

            if (root.data < x) { // check if root value is less than x
                fl = root.data; // update the floor value to root value
                root = root.right; // move to the right child
            } else {
                root = root.left; // move to the left child
            }
        }
        return fl; // return the floor value
    }
}

File: Floor in BST - GFG/floor-in-bst.js
//{ Driver Code Starts
//Initial Template for javascript

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

class Node {
    constructor(data) {
        this.data = data;
        this.right = null;
        this.left = null;
    }
}



function insert(tree, val) {
    if (tree === null) return new Node(val);
    
    if (val < tree.data) {
        tree.left = insert(tree.left, val);
    } else if (val > tree.data) {
        tree.right = insert(tree.right, val);
    }

    return tree;
}

function main() {
    const t = parseInt(readLine());
    for(let i = 0; i < t; i++) {
        let root = null;

        const n = parseInt(readLine());
        const elements = readLine().split(' ').map(Number);

        for (const el of elements) {
            root = insert(root, el);
        }

        const s = parseInt(readLine());
        
        console.log(floor(root, s));
    }
}



// } Driver Code Ends
//User function Template for javascript
function floor(root, x) {
    //code here
    let fl = -1;  // Variable to store the floor value, initialized as -1.

    // Traverse the BST until we reach a leaf node or find the target node.
    while (root) {
        if (root.data === x) {
            return root.data;  // If the current node value matches the target, return the target value.
        }

        if (root.data < x) {
            fl = root.data;  // If the current node value is less than the target, update the floor value.
            root = root.right;  // Move to the right subtree.
        } else {
            root = root.left;  // If the current node value is greater than or equal to the target, move to the left subtree.
        }
    }
    return fl;  // Return the floor value.
}

//{ Driver Code Starts.
// } Driver Code Ends